src/test/PoolSwapWithFeeTest.sol

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.20;
+
+import {CurrencyLibrary, Currency} from "../types/Currency.sol";
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+import {BalanceDelta, BalanceDeltaLibrary} from "../types/BalanceDelta.sol";
+import {PoolKey} from "../types/PoolKey.sol";
+import {IHooks} from "../interfaces/IHooks.sol";
+import {Hooks} from "../libraries/Hooks.sol";
+import {PoolTestBase} from "./PoolTestBase.sol";
+import {Test} from "forge-std/Test.sol";
+import {Hooks} from "../libraries/Hooks.sol";
+import {IHooks} from "../interfaces/IHooks.sol";
+import {SafeCast} from "../libraries/SafeCast.sol";
+
+contract PoolSwapWithFeeTest is Test, PoolTestBase {
+    using CurrencyLibrary for Currency;
+    using Hooks for IHooks;
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    constructor(IPoolManager _manager) PoolTestBase(_manager) {}
+
+    uint256 public constant TOTAL_DEBT = uint256(0);
+
+    error NoSwapOccurred();
+
+    struct CallbackData {
+        address sender;
+        PoolKey key;
+        IPoolManager.SwapParams params;
+        uint256 minAmountOut;
+    }
+
+    function swapExactInput(PoolKey memory key, IPoolManager.SwapParams memory params, uint256 minAmountOut)
+        external
+        payable
+        returns (BalanceDelta delta)
+    {
+        delta = abi.decode(
+            manager.lock(address(this), abi.encode(CallbackData(msg.sender, key, params, minAmountOut))), (BalanceDelta)
+        );
+
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) CurrencyLibrary.NATIVE.transfer(msg.sender, ethBalance);
+    }
+
+    function lockAcquired(address sender, bytes calldata rawData) external returns (bytes memory) {
+        require(msg.sender == address(manager));
+
+        CallbackData memory data = abi.decode(rawData, (CallbackData));
+        sender = (sender == address(this)) ? data.sender : sender;
+
+        BalanceDelta delta = manager.swap(data.key, data.params, "");
+
+        if (data.params.zeroForOne) {
+            // settle input
+            _settle(data.key.currency0, sender, delta.amount0(), true);
+
+            // settle output and fee
+            manager.acceptDebt(data.key.currency1, address(data.key.hooks), TOTAL_DEBT);
```

Comment:
i can get behind this pattern. almost all debt creation will originate from the hook address (hook fees, custom curves)

also it creates a "best practice" where debt _should_ originate from the hook address and not some external contract

---

an alternative would be implementing a way to pay down all debt (regardless of address) but I dont think this is feasible because of the mapping-based accounting

Reply:
yeah thats a nice point about it starting a best practice - i hadnt thought of that.

This function also makes it possible for the hook to "transfer" funds back to the router by having the hook call `manager.acceptDebt(router, x)`. The router will then have a balance of x tokens they can withdraw and give to the user. You can imagine this for a case where the hook performs a swap or withdraws liquidity and wants to give the output tokens to the user. The second use-case `acceptDebt` is called _before_ the debt is "created" so i think we're going to rename the function to `settleFor` or `payOnBehalf` or something like that.